build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java

```diff
@@ -0,0 +1,129 @@
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+import javax.inject.Inject;
+
+import static org.objectweb.asm.Opcodes.V_PREVIEW;
+
+public class MrjarPlugin implements Plugin<Project> {
+
+    private static final Pattern JAVA_SOURCESET_PATTERN = Pattern.compile("java(\\d{2})");
+
+    private final JavaToolchainService javaToolchains;
+
+    @Inject
+    MrjarPlugin(JavaToolchainService javaToolchains) {
+        this.javaToolchains = javaToolchains;
+    }
+
+    @Override
+    public void apply(Project project) {
+        project.getPluginManager().apply(JavaLibraryPlugin.class);
+        var javaExtension = project.getExtensions().getByType(JavaPluginExtension.class);
+
+        var srcDir = project.getProjectDir().toPath().resolve("src");
+        try (var subdirStream = Files.list(srcDir)) {
+            for (Path sourceset : subdirStream.toList()) {
+                String sourcesetName = sourceset.getFileName().toString();
+                Matcher sourcesetMatcher = JAVA_SOURCESET_PATTERN.matcher(sourcesetName);
+                if (sourcesetMatcher.matches()) {
+                    int javaVersion = Integer.parseInt(sourcesetMatcher.group(1));
+                    addMrjarSourceset(project, javaExtension, sourcesetName, javaVersion);
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void addMrjarSourceset(Project project, JavaPluginExtension javaExtension, String sourcesetName, int javaVersion) {
+        SourceSet sourceSet = javaExtension.getSourceSets().maybeCreate(sourcesetName);
+        GradleUtils.extendSourceSet(project, SourceSet.MAIN_SOURCE_SET_NAME, sourcesetName);
+
+        Jar jarTask = project.getTasks().withType(Jar.class).named(JavaPlugin.JAR_TASK_NAME).get();
+        jarTask.dependsOn(sourceSet.getCompileJavaTaskName());
+        jarTask.into("META-INF/versions/" + javaVersion, copySpec -> copySpec.from(sourceSet.getOutput()));
+        jarTask.manifest(manifest -> {
+            manifest.attributes(Map.of("Multi-Release", "true"));
+        });
+
+        JavaCompile compileTask = project.getTasks().withType(JavaCompile.class).named(sourceSet.getCompileJavaTaskName()).get();
+        compileTask.getJavaCompiler().set(javaToolchains.compilerFor(spec -> {
+            spec.getLanguageVersion().set(JavaLanguageVersion.of(javaVersion));
+        }));
+        compileTask.setSourceCompatibility(Integer.toString(javaVersion));
+        CompileOptions compileOptions = compileTask.getOptions();
+        compileOptions.getRelease().set(javaVersion);
+        compileOptions.getCompilerArgs().add("--enable-preview");
+        compileOptions.getCompilerArgs().add("-Xlint:-preview");
+
+        compileTask.doLast(t -> {
```

Comment:
Thinking about this more we should confirm this doesn't break things like cacheability. @breskby does adding custom actions to task have any other additional consequences that I'm unaware of?

Reply:
I ran a compile task after doing a clean, and it looks like it correctly pulled it from the cache:
https://gradle-enterprise.elastic.co/s/7j7ig5ljmf6ka